### ISSUE TITLE
No easy way to obtain row index from queries?

### ISSUE THREAD

**Issue by oceanusxiv** [lancedb/lance#3140]
Here's the use case, I wish to perform a random access `take` of a dataset, given some indices which was obtained by some query beforehand. Effectively I wish to implement my own sampler (I can't use the native sampler because I also want to do some index offsets for lookahead and behind).

However, there seems to be no easy way currently to do this. Primarily this seems to be due to an asymmetry in the Python API between the `take` function, which expects row indices, and all other query functions, which only return row ids, or row addresses.

I realize of course that there is a 1-1 mapping between row addresses, and row indices, but that mapping is hardly straightforward to calculate for the end user, and it would just be super convenient if we can have a `with_row_indices` option in all our query functions so we can obtain this information without so much hassle.

If I just missed something and such a method exists, do let me know :)

**Comment by wjones127** [comment]
You are right there isn't an easy way. Definitely something we could add. In the mean time, here is a snippet one could use to derive this:

```python
fragment_sizes = [(f.fragment_id, f.count_rows()) for f in ds.get_fragments()]
offsets = {}
offset = 0
for fragment_id, size in fragment_sizes:
    offsets[fragment_id] = offset
    offset += size

def row_addr_to_index(row_addr):
    fragment_id = row_addr >> 32
    row_offset = row_addr & 0xffffffff
    row_index = offsets[fragment_id] + row_offset
    return row_index

row_addrs = ds.to_table(with_row_address=True)['_rowaddr']
[row_addr_to_index(row_addr.as_py()) for row_addr in row_addrs]
```

**Comment by westonpace** [comment]
If we do end up adding such a feature at some point I would recommend calling it the "dataset offset" and not "row indices" as I think that is a little less ambiguous.

**Comment by westonpace** [comment]
Closed by https://github.com/lancedb/lance/pull/4478

Example (uses `lancedb` but should be similar in `lance`):

```
import lancedb
import pyarrow as pa

db = lancedb.connect("/tmp/foodb")
tbl = db.create_table("blah", pa.table({"x": range(100)}))
tbl.add(pa.table({"x": range(100, 200)}))

tbl.search(None).where("x > 95 AND x < 105").select(["x", "_rowoffset", "_rowid"]).to_arrow()
# pyarrow.Table
# x: int64
# _rowoffset: uint64
# _rowid: uint64
# ----
# x: [[96,97,98,99],[100,101,102,103,104]]
# _rowoffset: [[96,97,98,99],[100,101,102,103,104]]
# _rowid: [[96,97,98,99],[4294967296,4294967297,4294967298,4294967299,4294967300]]
```
